base/array.jl

```diff
@@ -2336,20 +2336,48 @@ Int64[]
 function findall(A)
     collect(first(p) for p in pairs(A) if last(p))
 end
+
 # Allocating result upfront is faster (possible only when collection can be iterated twice)
-function findall(A::AbstractArray{Bool})
-    n = count(A)
+function _findall(f::Function, A::AbstractArray{Bool})
+    # Compute f for true and false only once
+    ft, ff = f(true), f(false)
+    (ft | ff) || return Vector{eltype(keys(A))}()
+    (ft & ff) && return vec(Array(keys(A)))
+    n = let
+        c = count(A)
+        ft ? c : length(A) - c
+    end
     I = Vector{eltype(keys(A))}(undef, n)
+    _findall!(I, ff, A)
+end
+
+function _findall!(I::Vector, invert::Bool, A::AbstractArray{Bool})
     cnt = 1
-    for (i,a) in pairs(A)
-        if a
-            I[cnt] = i
-            cnt += 1
-        end
+    len = length(I)
+    for (k, v) in pairs(A)
+        cnt > len && break
+        I[cnt] = k
+        cnt += v ⊻ invert
     end
     I
 end
 
+function _findall!(I::Vector, invert::Bool, A::AbstractVector{Bool})
+    i = firstindex(A)
+    cnt = 1
+    len = length(I)
+    @inbounds while cnt ≤ len
```

Comment:
```suggestion
    while cnt ≤ len
```

Reply:
I still don't understand the problem here... :-/

Reply:
Just out of curiosity, I would also like to know the policy here. Is the idea that a custom subtype of AbstractArray may implement indexing wrong, and therefore it's better to throw an indexerror instead of a potential segfault, even when we know it's inbounds for well-defined arrays? The cost of boundschecking here is significant - about 60% increased runtime for long vectors on my laptop.